### PR TITLE
[Fixes #6160] Adds article for Write-Progress while multithreading

### LINKED ISSUE #6160: Foreach-Object -Parallel examples of how to report progress
It would be great if we could get more examples of some complex scenarios using the -Parallel parameter.  Recently I wanted to use Write-Progress, but as I had assumed, it doesn't work across runspaces.  I resorted to using the concurrent bag collection to add objects at the start of the runspace and then get the count of items and then over-write the same line with "x of n completed" sort of thing.  If there is a better way to do this I would love to see some examples.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: da56c3ce-409c-dd44-888b-6a875e93d4a4
* Version Independent ID: 7ba53b91-ed64-cc24-0cf0-a5dc3f3dc9d0
* Content: [ForEach-Object (Microsoft.PowerShell.Core)](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/foreach-object?view=powershell-7.1)
* Content Source: [reference/7.1/Microsoft.PowerShell.Core/ForEach-Object.md](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/7.1/Microsoft.PowerShell.Core/ForEach-Object.md)
* Product: **powershell**
* GitHub Login: @SteveL-MSFT
* Microsoft Alias: **slee**

### PR DESCRIPTION
**Pull request by chasewilson**
# PR Summary

There has been no article for how to write progress when using multithreading. 

## PR Context

Fixes #6160 
Fixes [AB#1753592](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1753592)

Select the area of the Table of Contents containing the documents being changed.

**Conceptual content**
- [ ] Overview and Install
- [x] Learning PowerShell
  - [ ] PowerShell 101
  - [x] Deep dives
  - [ ] Remoting
- [ ] Release notes (What's New)
- [ ] Windows PowerShell
  - WMF, ISE, release notes, etc.
- [ ] DSC articles
- [ ] Community resources
- [ ] Sample scripts
- [ ] Gallery articles
- [ ] Scripting and development
  - [ ] Legacy SDK

**Cmdlet reference & about_ topics**
- [ ] Version 7.1 preview content
- [ ] Version 7.0 content
- [ ] Version 6 content
- [ ] Version 5.1 content

## PR Checklist

- [x] I have read the [contributors guide][contrib] and followed the style and process guidelines
- [x] PR has a meaningful title
- [x] PR is targeted at the _staging_ branch
- [x] All relevant versions updated
- [x] Includes content related to issues and PRs - see [Closing issues using keywords][key].
- [x] This PR is ready to merge and is not **Work in Progress**
  - If the PR is work in progress, please add the prefix `WIP:` or `[WIP]` to the beginning of the
    title and remove the prefix when the PR is ready.

[contrib]: https://docs.microsoft.com/powershell/scripting/community/contributing/overview
[key]: https://help.github.com/en/articles/closing-issues-using-keywords

### CODE REVIEW

File: reference/docs-conceptual/learn/deep-dives/write-progress-across-multiple-threads.md [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+monitor the progress of a process using [Write-Progress](/powershell/reference/7.0/Microsoft.PowerShell.Utility/Write-Progress).
+However, since PowerShell uses a separate runspace for each thread when using **Parallel**,
+reporting the progress back to the host isn't as straight forward as normal use of `Write-Progress`.
+
+## Using a synced hashtable to track progress
+
+When writing the progress from multiple threads, tracking becomes difficult because when running
+parallel processes in PowerShell, each process has it's own runspace. To get around this, you can
+use a [synchronized hashtable](/dotnet/api/system.collections.hashtable.synchronized). A synced
+hashtable is a thread safe data structure that can be edited by multiple threads simultaneously
```
Comment (sdwheeler) [inline reference/docs-conceptual/learn/deep-dives/write-progress-across-multiple-threads.md:21]:
```suggestion
hashtable is a thread safe data structure that can be modified by multiple threads simultaneously
```

File: reference/docs-conceptual/learn/deep-dives/write-progress-across-multiple-threads.md [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+```
+
+This section creates three different data structures, for three different purposes.
+
+The `$dataSet` variable stores an array of hashtables that is used to coordinate the next steps
+without the risk of being modified. When looping through an object in PowerShell, if the object
+being looped through is modified, PowerShell will throw an error. So, keeping the object being
+looped through, separate from the objects being modified is a must. The `Id` key in each hashtable
+is the identifier for a mock process. The `Wait` key simulates the workload of each mock process to
+track.
```
Comment (sdwheeler) [inline reference/docs-conceptual/learn/deep-dives/write-progress-across-multiple-threads.md:69]:
```suggestion
The `$dataSet` variable stores an array of hashtables that is used to coordinate the next steps
without the risk of being modified. If a object collection is modified while iterating through the
collection, PowerShell throws an error. You must keep the object collection in the loop separate
from the objects being modified. The `Id` key in each hashtable is the identifier for a mock
process. The `Wait` key simulates the workload of each mock process being tracked.
```

File: reference/docs-conceptual/learn/deep-dives/write-progress-across-multiple-threads.md [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+This section runs the multi-threaded processes and creates some of the output used to display
+progress.
+
+```powershell
+$job = $dataset | Foreach-Object -ThrottleLimit 3 -AsJob -Parallel {
+    $syncCopy = $using:sync
+    $process = $syncCopy.$($PSItem.Id)
+
+    $process.Id = $PSItem.Id
+    $process.Activity = "$($PSItem.Id) starting"
```
Comment (sdwheeler) [inline reference/docs-conceptual/learn/deep-dives/write-progress-across-multiple-threads.md:87]:
```suggestion
    $process.Activity = "Id $($PSItem.Id) starting"
```

File: reference/docs-conceptual/learn/deep-dives/write-progress-across-multiple-threads.md [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    $process.Status = "Processing"
+
+    # Fake workload start up that takes x amount of time to complete
+    start-sleep -Milliseconds ($PSItem.wait*5)
+
+    # Process. update activity
+    $process.Activity = "$($PSItem.id) processing"
```
Comment (sdwheeler) [inline reference/docs-conceptual/learn/deep-dives/write-progress-across-multiple-threads.md:94]:
```suggestion
    $process.Activity = "Id $($PSItem.id) processing"
```

File: reference/docs-conceptual/learn/deep-dives/write-progress-across-multiple-threads.md [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    }
+
+    # Wait to refresh to not overload gui
+    Start-Sleep -Seconds 0.1
+}
+```
+
+The `$job` variable contains the parent **job** and has a child **job** for each of the mock
+processes. While any of the child jobs are still running, the parent job **State** will remain
+"Running". This allows us to reliable use the `while` loop to continually update the progress of
```
Comment (sdwheeler) [inline reference/docs-conceptual/learn/deep-dives/write-progress-across-multiple-threads.md:156]:
```suggestion
"Running". This allows us to use the `while` loop to continually update the progress of
```

File: reference/docs-conceptual/learn/deep-dives/write-progress-across-multiple-threads.md [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+
+# Create synced hashtable
+$sync = [System.Collections.Hashtable]::Synchronized($origin)
+
+$job = $dataset | Foreach-Object -ThrottleLimit 3 -AsJob -Parallel {
+    $syncCopy = $using:sync
+    $process = $syncCopy.$($PSItem.Id)
+
+    $process.Id = $PSItem.Id
+    $process.Activity = "$($PSItem.Id) starting"
```
Comment (sdwheeler) [inline reference/docs-conceptual/learn/deep-dives/write-progress-across-multiple-threads.md:208]:
```suggestion
    $process.Activity = "Id $($PSItem.Id) starting"
```

File: reference/docs-conceptual/learn/deep-dives/write-progress-across-multiple-threads.md [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    $process.Status = "Processing"
+
+    # Fake workload start up that takes x amount of time to complete
+    start-sleep -Milliseconds ($PSItem.wait*5)
+
+    # Process. update activity
+    $process.Activity = "$($PSItem.id) processing"
```
Comment (sdwheeler) [inline reference/docs-conceptual/learn/deep-dives/write-progress-across-multiple-threads.md:215]:
```suggestion
    $process.Activity = "Id $($PSItem.id) processing"
```

**Comment by sdwheeler** [comment]
@WithHolm Can you review the PR?

**Comment by WithHolm** [comment]
Looks good to me!

**Comment by chasewilson** [comment]
@WithHolm Thank you for the example. It's a great contribution and will greatly help the community!

**Pull request merged**